### PR TITLE
Expose get/set deployment config in SQL CLI

### PR DESCRIPTION
**Pull request by tatiana**
Refactor the retrieval of configuration to allow retrieving most of the settings by exposing them as JSON.
```
$ flow config get airflow_home
$ flow config get --as-json
```

Add command to set deployment configuration per environment, using:
```
$ flow config set deploy --env=dev --astro-workspace=cl123 --astro-deplyment-cl345
```

As part of this PR, I also refactored the Airflow home to be environment-specific.

Disclaimers:
1. The command config is hidden from end-users and is meant to be used to integrate with the Astro CLI.
2. We'll need a follow up ticket on the the Astro CLI side, changing the current call from ` flow config airflow_home` to  flow config get airflow_home`

Related: #1166

### CODE REVIEW

File: sql-cli/tests/test_configuration.py [hunk 1]
```diff
@@ -19,5 +19,4 @@ def test_from_yaml_to_config_without_env():
     config_from_file = config_reference.from_yaml_to_config()
     assert isinstance(config_from_file, Config)
     assert config_from_file.project_dir == config_reference.project_dir
-    assert config_from_file.environment is None
-    assert config_from_file.connections == []
+    assert config_from_file.environment == "default"
```
Comment (tatiana) [inline sql-cli/tests/test_configuration.py:22]: @feluelle, I'd love your thoughts on this change.

My feeling is that it is valid in the use case where we use the Configuration object and we don't care about the environment - but I thought it might be more graceful to have a default value than to have to propagate str/None possibilities through the code, handling scenarios where config would not have an environment.

File: sql-cli/include/base/config/dev/configuration.yml [hunk 2]
```diff
@@ -9,3 +9,7 @@ connections:
   - conn_id: sqlite_conn
     conn_type: sqlite
     host: retail.db
+
+# Apache Airflow optional settings
+airflow:
+  home:
```
Comment (tatiana) [inline sql-cli/include/base/config/dev/configuration.yml:15]: @pankajkoti I'd love your thoughts on this change, since you worked a lot on this part of the code as well

File: sql-cli/tests/test_configuration.py (hunk 1 above)
Comment (dimberman) [reply sql-cli/tests/test_configuration.py:22]: I personally think this makes sense. Default is... well... a default so it might as well be reflected as such in the code.

Comment (feluelle) [reply sql-cli/tests/test_configuration.py:22]: Yes, +1 on explicit `default` as a default environment.

File: sql-cli/sql_cli/constants.py [hunk 3]
```diff
@@ -3,7 +3,7 @@
 CONFIG_FILENAME = "configuration.yml"
 
 DEFAULT_ENVIRONMENT = "default"
-DEFAULT_AIRFLOW_HOME = f".airflow/{DEFAULT_ENVIRONMENT}"
+DEFAULT_BASE_AIRFLOW_HOME = ".airflow/"
```
Comment (feluelle) [inline sql-cli/sql_cli/constants.py:6]:
```suggestion
DEFAULT_BASE_AIRFLOW_HOME = ".airflow"
```
should work too, and makes it consistent with the other paths.

File: sql-cli/sql_cli/project.py [hunk 4]
```diff
@@ -80,51 +83,85 @@ def data_dir(self) -> Path:
...
+
+        :param airflow_home: Path to Airflow home we want to clean up.
         """
-        logs_folder = self.airflow_home / "logs"
+        logs_folder = airflow_home / "logs"
         shutil.rmtree(logs_folder)
 
-        webserver_config = self.airflow_home / "webserver_config.py"
+        webserver_config = airflow_home / "webserver_config.py"
         webserver_config.unlink()
```
Comment (feluelle) [inline sql-cli/sql_cli/project.py:153]: This function should actually live in our airflow utils and maybe get called in `initialise_airflow` at the end. We get the airflow_home there as well as an arg. WDYT?

File: sql-cli/sql_cli/project.py [hunk 5]
```diff
@@ -145,8 +182,9 @@ def initialise(self) -> None:
             dirs_exist_ok=True,
         )
         self._initialise_global_config()
-        initialise_airflow(self.airflow_home, self.airflow_dags_folder)
-        self._remove_unnecessary_airflow_files()
+        for env in self.environments_list:
+            self._initialise_env_config(env)
+            self._initialise_airflow_for_environment(env)
```
Comment (feluelle) [inline sql-cli/sql_cli/project.py:187]: Nice!

File: sql-cli/sql_cli/project.py [hunk 6]
```diff
@@ -168,11 +206,18 @@ def load_config(self, environment: str = DEFAULT_ENVIRONMENT) -> None:
         config = Config(environment=environment, project_dir=self.directory).from_yaml_to_config()
         if config.airflow_home:
             self._airflow_home = Path(config.airflow_home).resolve()
+        else:
+            self._initialise_env_config(environment)
+            config = Config(environment=environment, project_dir=self.directory).from_yaml_to_config()
+
+        if not self.get_env_airflow_home(environment).exists():
+            self._initialise_airflow_for_environment(environment)
+
```
Comment (feluelle) [inline sql-cli/sql_cli/project.py:215]: Why did we add this? This should be done by `initialise`.

File: sql-cli/sql_cli/project.py [hunk 7]
```diff
@@ -168,11 +206,18 @@ def load_config(self, environment: str = DEFAULT_ENVIRONMENT) -> None:
...
         if config.airflow_dags_folder:
             self._airflow_dags_folder = Path(config.airflow_dags_folder).resolve()
         if config.data_dir:
             self._data_dir = Path(config.data_dir).resolve()
-        reload_airflow(self.airflow_home)
+        reload_airflow(self.get_env_airflow_home(environment))
```
Comment (feluelle) [inline sql-cli/sql_cli/project.py:220]:
```suggestion
        reload_airflow(self._airflow_home)
```
?

File: sql-cli/tests/config/test/configuration.yml [hunk 8]
```diff
@@ -22,7 +22,7 @@ connections:
     schema: pagila
     login: postgres
     password: $POSTGRES_PASSWORD
-    port: $POSTGRES_PORT
+    port: 5432
```
Comment (feluelle) [inline sql-cli/tests/config/test/configuration.yml:25]: Please revert this change. You can set the env vars in your `.env` file.

File: sql-cli/sql_cli/cli/config.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+    cls=AstroCommand,
+    help="""
+   Set the project configuration.
+
+   Example:
+   $ flow config set deploy --env=dev --astro-workspace=cl123 --astro-deployment=cl345
+   """,
+)
+# skipcq: PYL-W0622
+def set(  # noqa: A001
```
[the remark this answers is not in the document]
Comment (feluelle) [reply sql-cli/sql_cli/cli/config.py:107]: Note that it exposes it as `set` see line 98 in the suggestion.

File: sql-cli/sql_cli/project.py (hunk 6 above)
Comment (tatiana) [reply sql-cli/sql_cli/project.py:215]: The following use case fails if we don't handle this afterwards:
```
$ flow init .
# user manually adds a new environment `stage` to the `config` folder
$ flow validate --env stage
```

I realised this use case with one of the existing tests!

Comment (feluelle) [reply sql-cli/sql_cli/project.py:215]: Right! Thanks 👍 Can we simplify this?

File: sql-cli/sql_cli/cli/config.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+    Get the project configuration.
+
+    Example of usages:
+    $ flow config get airflow_home
+    $ flow config get --as-json
+
+    The first returns a key from the config whereas the second returns all the configuration as JSON.
+    """,
+)
+def get(
```
[the remark this answers is not in the document]
Comment (tatiana) [reply sql-cli/sql_cli/cli/config.py:78]: Much better, changing it!

File: sql-cli/sql_cli/cli/config.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+
+def _set(key: str, project_dir: Path, env: str, astro_deployment: str, astro_workspace: str) -> None:
+    """
+    Set deployment configuration associated to a SQL CLI environment.
+
+    :param key: Configuration property (key) to be set. At the mmoment only "deploy" is accepted.
+    :param project_dir: Path to the project directory.
+    :param env: SQL CLI environment (default, dev).
+    :param astro_deployment: Astronomer Cloud deployment ID.
```
Comment (pankajkoti) [inline sql-cli/sql_cli/cli/config.py:48]: Should we rename this to `astro_deployment_id` ?

File: sql-cli/sql_cli/cli/config.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+    :param astro_workspace: Astronomer Cloud deployment workspace.
```
Comment (pankajkoti) [inline sql-cli/sql_cli/cli/config.py:49]: Will this be workspace ID?

File: sql-cli/sql_cli/cli/config.py [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+    ),
+    env: str = typer.Option(
+        default=DEFAULT_ENVIRONMENT,
+        help="(Optional) Environment used to fetch the configuration key from.",
+    ),
+    as_json: bool = typer.Option(
+        False, "--json", help="If the response should be in JSON format", show_default=True
+    ),
+) -> None:
+    value = _get(key, project_dir, env, as_json)
```
Comment (pankajkoti) [inline sql-cli/sql_cli/cli/config.py:96]: Should we check that both `key` and `as_json` are not set together?

File: sql-cli/sql_cli/configuration.py [hunk 14]
```diff
@@ -94,3 +92,16 @@ def write_value_to_yaml(self, section: str, key: str, value: str, filepath: Path
...
+        """
+        Return configuration represented as JSON string.
+
+        :returns: Dictionary containing non-sensitive global and per-environment configuration.
+        """
+        config_as_dict = {"global": self.from_yaml_to_dict(self.get_global_config_filepath())}
+        if self.environment:
+            env_config = self.from_yaml_to_dict(self.get_env_config_filepath())
+            env_config.pop("connections", None)  # remove sensitive information
+            config_as_dict[self.environment] = env_config
```
Comment (pankajkoti) [inline sql-cli/sql_cli/configuration.py:106]:
```suggestion
            config_as_dict["local"] = env_config
```
Would it make sense to rename the key to "local", or the requestor will know how exactly to parse the response for environment values?

File: sql-cli/sql_cli/project.py [hunk 15]
```diff
@@ -80,50 +87,71 @@ def data_dir(self) -> Path:
...
 
     def _initialise_global_config(self) -> None:
         """
-        Initialises global config file that includes configuration to be shared across environments including the
+        Initialises global config YAML file that includes configuration to be shared across environments including the
         airflow config.
         """
-        config = Config(project_dir=self.directory)
-        global_env_filepath = config.get_global_config_filepath()
+        config = Config(environment=DEFAULT_ENVIRONMENT, project_dir=self.directory)
```
Comment (pankajkoti) [inline sql-cli/sql_cli/project.py:116]: @feluelle had reservations on passing Default environment for initialising global config. His question was why do we need to pass environment for global config, which I also felt is a valid question.

File: sql-cli/sql_cli/project.py [hunk 16]
```diff
@@ -145,8 +173,9 @@ def initialise(self) -> None:
             dirs_exist_ok=True,
         )
         self._initialise_global_config()
-        initialise_airflow(self.airflow_home, self.airflow_dags_folder)
-        self._remove_unnecessary_airflow_files()
+        for env in self.environments_list:
+            self._initialise_env_config(env)
+            self._initialise_airflow_for_environment(env)
```
Comment (pankajkoti) [inline sql-cli/sql_cli/project.py:178]: How will we take care of initialising Airflow when a user adds a new environment?

Comment (pankajkoti) [reply sql-cli/sql_cli/project.py:178]: I am thinking if we should provide a command to create a new environment which would take care of it.

File: sql-cli/sql_cli/project.py (hunk 15 above)
Comment (pankajkoti) [inline sql-cli/sql_cli/project.py:116]:
```suggestion
        config = Config(project_dir=self.directory)
```
Why not just this since Config now already has environment = DEFAULT_ENVIRONMENT as the default value?

File: sql-cli/sql_cli/project.py (hunk 6 above)
Comment (tatiana) [reply sql-cli/sql_cli/project.py]: @feluelle I added comments - how would you suggest simplifying it?

File: sql-cli/sql_cli/project.py (hunk 7 above)
Comment (tatiana) [reply sql-cli/sql_cli/project.py:219]: They are not the same thing, `self._airflow_home` is used during setup if the user explicitly specifies it

File: sql-cli/sql_cli/configuration.py (hunk 14 above)
Comment (tatiana) [reply sql-cli/sql_cli/configuration.py:106]: @pankajkoti, at the moment, it returns for all environments - and I'd expect the one using the endpoint to know which environment they are interested in. Do you see challenges with this approach?

File: sql-cli/sql_cli/project.py (hunk 16 above)
Comment (tatiana) [reply sql-cli/sql_cli/project.py:178]: >  How will we take care of initialising Airflow when a user adds a new environment?
@pankajkoti that is being handled in the `load_config`

In future we can look into exposing creating a new environment in a separate command, but it feels we don't need this ATM

File: sql-cli/sql_cli/project.py (hunk 15 above)
Comment (tatiana) [reply sql-cli/sql_cli/project.py:116]: I cleaned it up - by the time I made this change, that was not the default :)

Comment (tatiana) [reply sql-cli/sql_cli/project.py:116]: Since I set the default value in `Config`, this is now possible :)

File: sql-cli/sql_cli/cli/config.py (hunk 11 above)
Comment (tatiana) [reply sql-cli/sql_cli/cli/config.py:52]: :+1: I'm fine with it, @pankajkoti , it would make it more consistent with the Astro CLI
https://github.com/astronomer/docs/blob/main/astro/cli/astro-deploy.md

File: sql-cli/sql_cli/cli/config.py (hunk 12 above)
Comment (tatiana) [reply sql-cli/sql_cli/cli/config.py:53]: Same as before - changing it

**Review by pankajkoti — APPROVED** [review]

**Review by feluelle — APPROVED** [review]
